### PR TITLE
Allow changing the anisotropic filter level at run-time per Viewport

### LINKED ISSUE #4673: Allow changing the anisotropic filter level at run-time
*Related to https://github.com/godotengine/godot-proposals/issues/3134.*

### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

The anisotropic filter level used in material rendering cannot be changed while the project is running. This can be too limiting for graphics settings menus, which generally allow changing the anisotropic filtering level to improve performance on low-end GPUs (or improve quality on high-end GPUs with memory bandwidth to spare).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allow changing anisotropic filter level at run-time. This includes setting it at 1× to disable anisotropic filtering on all materials within the project, even if their sampling mode is set to anisotropic.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Similar to what https://github.com/godotengine/godot/pull/51679 did with mipmap LOD bias, the max anisotropy should be separated from the sampler state creation process so it can be adjusted at run-time.

Max anisotropy is currently defined here: https://github.com/godotengine/godot/blob/68b86220c806fa02d0d2b21d6561e7ea871e840e/servers/rendering/renderer_rd/storage_rd/material_storage.cpp#L1338-L1388

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Anisotropic filtering can be disabled on a per-material basis (per-texture basis in `3.x`), but this requires a lot of code to consistently apply throughout an entire project at run-time. Also, this does not allow using a lower or higher level of anisotropic filtering than the default while leaving it enabled.

### Is there a reason why this should be core and not an add-on in the asset library?

Anisotropic filtering level can't be overridden by an add-on.

### PR DESCRIPTION
**Pull request by wagnerfs**
Fixes https://github.com/godotengine/godot-proposals/issues/4673

Changes to anisotropic filtering on project settings are real time now and don't need an engine restart, also adds `anisotropic_filtering_level` property to viewports and can be changed separately.

Added the additional lines do the docs on viewports as well, I'm not sure what to do on the `ProjectSettings.xml` though besides removing the note about not being able to change at run-time.

A new enum had to be added though similar to MSAA
```c++
enum AnisotropicFiltering {
      ANISOTROPY_DISABLED,
      ANISOTROPY_2X,
      ANISOTROPY_4X,
      ANISOTROPY_8X,
      ANISOTROPY_16X,
      ANISOTROPY_MAX
};
```

### CODE REVIEW

File: doc/classes/RenderingServer.xml [hunk 1]
```diff
@@ -3776,6 +3776,16 @@
 				If [code]true[/code], canvas item vertices (i.e. polygon points) are snapped to the nearest pixel when rendering. This can lead to a crisper appearance at the cost of less smooth movement, especially when [Camera2D] smoothing is enabled. Equivalent to [member ProjectSettings.rendering/2d/snap/snap_2d_vertices_to_pixel].
 			</description>
 		</method>
+		<method name="viewport_set_anisotropic_filtering_level">
+			<return type="void" />
+			<param index="0" name="viewport" type="RID" />
+			<param index="1" name="anisotropic_filtering_level" type="int" />
+			<description>
+				Sets the maximum number of samples to take when using anisotropic filtering on textures (as a power of two). A higher sample count will result in sharper textures at oblique angles, but is more expensive to compute. A value of [code]0[/code] forcibly disables anisotropic filtering, even on materials where it is enabled.
+				The anisotropic filtering level also affects decals and light projectors if they are configured to use anisotropic filtering. See [member rendering/textures/decals/filter] and [member rendering/textures/light_projectors/filter].
```
Comment (AThousandShips) [inline doc/classes/RenderingServer.xml:3785]:
```suggestion
				The anisotropic filtering level also affects decals and light projectors if they are configured to use anisotropic filtering. See [member ProjectSettings.rendering/textures/decals/filter] and [member ProjectSettings.rendering/textures/light_projectors/filter].
```

File: doc/classes/Viewport.xml [hunk 2]
```diff
@@ -362,6 +362,11 @@
 			[b]Note:[/b] If [member scaling_3d_scale] is lower than [code]1.0[/code] (exclusive), [member texture_mipmap_bias] is used to adjust the automatic mipmap bias which is calculated internally based on the scale factor. The formula for this is [code]log2(scaling_3d_scale) + mipmap_bias[/code].
 			To control this property on the root viewport, set the [member ProjectSettings.rendering/textures/default_filters/texture_mipmap_bias] project setting.
 		</member>
+		<member name="anisotropic_filtering_level" type="int" setter="set_anisotropic_filtering_level" getter="get_anisotropic_filtering_level" default="2">
+			Sets the maximum number of samples to take when using anisotropic filtering on textures (as a power of two). A higher sample count will result in sharper textures at oblique angles, but is more expensive to compute. A value of [code]0[/code] forcibly disables anisotropic filtering, even on materials where it is enabled.
+			The anisotropic filtering level also affects decals and light projectors if they are configured to use anisotropic filtering. See [member rendering/textures/decals/filter] and [member rendering/textures/light_projectors/filter].
```
Comment (AThousandShips) [inline doc/classes/Viewport.xml:367]:
```suggestion
			The anisotropic filtering level also affects decals and light projectors if they are configured to use anisotropic filtering. See [member ProjectSettings.rendering/textures/decals/filter] and [member ProjectSettings.rendering/textures/light_projectors/filter].
```

**Comment by DarioSamo** [comment]
I can't speak for the GLES3 changes but the RD ones look good to me, glad to see the effort of splitting the samplers into being per-viewport was continued like this. The change makes sense to me.

File: drivers/gles3/storage/render_scene_buffers_gles3.cpp [hunk 3]
```diff
@@ -131,6 +131,7 @@ void RenderSceneBuffersGLES3::configure(const RenderSceneBuffersConfiguration *p
 	scaling_3d_mode = p_config->get_scaling_3d_mode();
 	//fsr_sharpness = p_config->get_fsr_sharpness();
 	//texture_mipmap_bias = p_config->get_texture_mipmap_bias();
+	//anisotropic_filtering_level = p_config->get_anisotropic_filtering_level();
```
Comment (DarioSamo) [inline drivers/gles3/storage/render_scene_buffers_gles3.cpp:134]: Is there a purpose to the commented code being introduced here?

File: servers/rendering/storage/render_scene_buffers.cpp [hunk 4]
```diff
@@ -61,11 +61,15 @@ void RenderSceneBuffersConfiguration::_bind_methods() {
 
 	ClassDB::bind_method(D_METHOD("get_fsr_sharpness"), &RenderSceneBuffersConfiguration::get_fsr_sharpness);
 	ClassDB::bind_method(D_METHOD("set_fsr_sharpness", "fsr_sharpness"), &RenderSceneBuffersConfiguration::set_fsr_sharpness);
-	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "fsr_sharpness"), "set_fsr_sharpness", "get_fsr_sharpness");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "fsr_sharpness"), "set_fsr_sharpness", "get_fsr_sharpness");
```
Comment (DarioSamo) [inline servers/rendering/storage/render_scene_buffers.cpp:64]: This is an interesting error, but I wouldn't have realized it it was part of this PR if I didn't inspect it. Perhaps it would be best to have this in its own PR so it can be reviewed independently what it was doing?

File: servers/rendering/storage/render_scene_buffers.cpp [hunk 5]
```diff
@@ -61,11 +61,15 @@ void RenderSceneBuffersConfiguration::_bind_methods() {
...
 
 	ClassDB::bind_method(D_METHOD("get_texture_mipmap_bias"), &RenderSceneBuffersConfiguration::get_texture_mipmap_bias);
 	ClassDB::bind_method(D_METHOD("set_texture_mipmap_bias", "texture_mipmap_bias"), &RenderSceneBuffersConfiguration::set_texture_mipmap_bias);
-	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "texture_mipmap_bias"), "set_texture_mipmap_bias", "get_texture_mipmap_bias");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "texture_mipmap_bias"), "set_texture_mipmap_bias", "get_texture_mipmap_bias");
```
Comment (DarioSamo) [inline servers/rendering/storage/render_scene_buffers.cpp:68]: Same as the previous comment.

File: doc/classes/Viewport.xml [hunk 6]
```diff
@@ -362,6 +362,11 @@
...
+			[b]Note:[/b] For performance reasons, anisotropic filtering [i]is not enabled by default[/i] on 2D and 3D materials. For this setting to have an effect in 3D, set [member BaseMaterial3D.texture_filter] to [constant BaseMaterial3D.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant BaseMaterial3D.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on materials. For this setting to have an effect in 2D, set [member CanvasItem.texture_filter] to [constant CanvasItem.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant CanvasItem.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on the [CanvasItem] node displaying the texture (or in [CanvasTexture]). However, anisotropic filtering is rarely useful in 2D, so only enable it for textures in 2D if it makes a meaningful visual difference.
```
Comment (Mickeon) [inline doc/classes/Viewport.xml:368]: > However, anisotropic filtering is rarely useful in 2D, so only enable it for textures in 2D if it makes a meaningful visual difference.

While I do agree with the message, the user should be told why this is the case instead of being vaguely "useless". I myself am not sure on how anisotropic filtering works in 2D. Going into a bit more detail would be nice.

On that note, this note is **looong**, and there is a notable distinction between 2D and 3D. Users may care about only one or both, so I would suggest splitting it in two:

```suggestion
			[b]Note:[/b] In 3D, for this setting to have an effect, set [member BaseMaterial3D.texture_filter] to [constant BaseMaterial3D.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant BaseMaterial3D.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on materials.
			[b]Note:[/b] In 2D, for this setting to have an effect, set [member CanvasItem.texture_filter] to [constant CanvasItem.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant CanvasItem.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on the [CanvasItem] node displaying the texture (or in [CanvasTexture]). However, anisotropic filtering is rarely useful in 2D, so only enable it for textures in 2D if it makes a meaningful visual difference.
```

File: servers/rendering/storage/render_scene_buffers.cpp (hunk 4 above)
Comment (wagnerfs) [reply servers/rendering/storage/render_scene_buffers.cpp:64]: Might've slid in since I'm using these changes on my own custom godot build, just seemed to me while changing the files that someone might accidentally left those as `BOOL`. I'll remove those and maybe make a new PR just for that, although I have no clue what they really do.

Comment (DarioSamo) [reply servers/rendering/storage/render_scene_buffers.cpp]: I think it sounds like a bug as well, the parameter is clearly meant to be a float. But I'd definitely make it part of a separate PR to track it down better and not have to associate that fix to this feature upgrade in case we want to cherry-pick or do regression testing.

File: doc/classes/RenderingServer.xml [hunk 7]
```diff
@@ -4676,6 +4676,24 @@
...
+			Use 2× Anisotropic Filtering.
+		</constant>
+		<constant name="VIEWPORT_ANISOTROPY_4X" value="2" enum="ViewportAnisotropicFiltering">
+			Use 4× Anisotropic Filtering. This is the default value.
+		</constant>
+		<constant name="VIEWPORT_ANISOTROPY_8X" value="3" enum="ViewportAnisotropicFiltering">
+			Use 8× Anisotropic Filtering.
+		</constant>
+		<constant name="VIEWPORT_ANISOTROPY_16X" value="4" enum="ViewportAnisotropicFiltering">
+			Use 8× Anisotropic Filtering.
```
Comment (Mickeon) [inline doc/classes/RenderingServer.xml:4692]: Caught you copy-pasting in 4K 📸
```suggestion
			Use 16× Anisotropic Filtering.
```

File: doc/classes/RenderingServer.xml [hunk 8]
```diff
@@ -4676,6 +4676,24 @@
 		<constant name="VIEWPORT_MSAA_MAX" value="4" enum="ViewportMSAA">
 			Represents the size of the [enum ViewportMSAA] enum.
 		</constant>
+		<constant name="VIEWPORT_ANISOTROPY_DISABLED" value="0" enum="ViewportAnisotropicFiltering">
+			Anisotropic filtering disabled.
```
Comment (Mickeon) [inline doc/classes/RenderingServer.xml:4680]:
```suggestion
			Anisotropic filtering is disabled.
```

File: servers/rendering/storage/render_scene_buffers.cpp (hunk 4 above)
Comment (Calinou) [reply servers/rendering/storage/render_scene_buffers.cpp]: The property hint type is just a suggestion for the inspector and class reference, but it should be fixed in a separate PR indeed.

File: drivers/gles3/storage/render_scene_buffers_gles3.cpp (hunk 3 above)
Comment (Calinou) [reply drivers/gles3/storage/render_scene_buffers_gles3.cpp:134]: While FSR1 probably won't be implemented in Compatibility, anisotropic filtering level change can be done on all platforms (including OpenGL ES) while mipmap LOD bias is limited to desktop OpenGL (it was never part of OpenGL ES).

File: doc/classes/ProjectSettings.xml [hunk 9]
```diff
@@ -2768,7 +2768,6 @@
...
 			The anisotropic filtering level also affects decals and light projectors if they are configured to use anisotropic filtering. See [member rendering/textures/decals/filter] and [member rendering/textures/light_projectors/filter].
 			[b]Note:[/b] For performance reasons, anisotropic filtering [i]is not enabled by default[/i] on 2D and 3D materials. For this setting to have an effect in 3D, set [member BaseMaterial3D.texture_filter] to [constant BaseMaterial3D.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant BaseMaterial3D.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on materials. For this setting to have an effect in 2D, set [member CanvasItem.texture_filter] to [constant CanvasItem.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant CanvasItem.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on the [CanvasItem] node displaying the texture (or in [CanvasTexture]). However, anisotropic filtering is rarely useful in 2D, so only enable it for textures in 2D if it makes a meaningful visual difference.
-			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this setting at run-time.
```
Comment (Calinou) [inline doc/classes/ProjectSettings.xml:2771]: Readd a note describing which property should be changed:

```
[b]Note:[/b] This property is only read when the project starts. To change the anisotropic filtering level at runtime, set [member Viewport.anisotropic_filter_level] on the root [Viewport] instead.
```

**Review by Calinou — COMMENTED** [review]
Tested locally (rebased on top of `master` 907db8eebcecb97d527edcaff77a1c87a6c068f5), it works as expected.

**Testing project:** [test_pr_88313.zip](https://github.com/godotengine/godot/files/14284820/test_pr_88313.zip)

https://github.com/godotengine/godot/assets/180032/8da6f732-2a6c-4779-a693-fe2d0fdc784e

**Review by Calinou — APPROVED** [review]
Code looks good to me. Please [squash commits together](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase) into a single commit so this can be merged.

**Comment by AThousandShips** [comment]
Please clean up your commit message, you don't need all the original messages, just a short message describing what's done

**Comment by wagnerfs** [comment]
Should be good to go now!

**Comment by akien-mga** [comment]
You'll need to run `godot --doctool`, or apply the diff from https://github.com/godotengine/godot/actions/runs/7906991378/job/21584760683?pr=88313 manually, to have the new properties and methods alphabetically sorted in the docs.

File: doc/classes/RenderingServer.xml [hunk 10]
```diff
@@ -4666,6 +4676,24 @@
 		<constant name="VIEWPORT_MSAA_MAX" value="4" enum="ViewportMSAA">
 			Represents the size of the [enum ViewportMSAA] enum.
 		</constant>
+		<constant name="VIEWPORT_ANISOTROPY_DISABLED" value="0" enum="ViewportAnisotropicFiltering">
+			Anisotropic filtering is disabled.
+		</constant>
+		<constant name="VIEWPORT_ANISOTROPY_2X" value="1" enum="ViewportAnisotropicFiltering">
+			Use 2× Anisotropic Filtering.
```
Comment (Mickeon) [inline doc/classes/RenderingServer.xml:4683]: While I did mention that it's best if "Filtering" is lowercase, I was not specific enough and did not mention that it should be like all lowercase for all **ViewportAnisotropicFiltering** and **AnisotropicFiltering** enum values. Sorry
```suggestion
			Use 2× anisotropic filtering.
```

**Comment by wagnerfs** [comment]
@akien-mga never dealt with docs before, thanks for the help

File: doc/classes/RenderingServer.xml [hunk 11]
```diff
@@ -3536,6 +3536,16 @@
...
+				The anisotropic filtering level also affects decals and light projectors if they are configured to use anisotropic filtering. See [member ProjectSettings.rendering/textures/decals/filter] and [member ProjectSettings.rendering/textures/light_projectors/filter].
+				[b]Note:[/b] For performance reasons, anisotropic filtering [i]is not enabled by default[/i] on 2D and 3D materials. For this setting to have an effect in 3D, set [member BaseMaterial3D.texture_filter] to [constant BaseMaterial3D.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant BaseMaterial3D.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on materials. For this setting to have an effect in 2D, set [member CanvasItem.texture_filter] to [constant CanvasItem.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant CanvasItem.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on the [CanvasItem] node displaying the texture (or in [CanvasTexture]). However, anisotropic filtering is rarely useful in 2D, so only enable it for textures in 2D if it makes a meaningful visual difference.
```
Comment (Mickeon) [inline doc/classes/RenderingServer.xml:3546]: This note was not split on Viewport but not here (you may want to just copy and paste it).

File: doc/classes/ProjectSettings.xml [hunk 12]
```diff
@@ -2768,7 +2768,7 @@
...
+			[b]Note:[/b] This property is only read when the project starts. To change the anisotropic filtering level at runtime, set [member Viewport.anisotropic_filtering_level] on the root [Viewport] instead.
```
Comment (Mickeon) [inline doc/classes/ProjectSettings.xml:2771]: I don't see why saying "instead" here. Not sure how to explain it.
```suggestion
			[b]Note:[/b] This property is only read when the project starts. To change the anisotropic filtering level at runtime, set [member Viewport.anisotropic_filtering_level] on the root [Viewport].
```

File: doc/classes/ProjectSettings.xml [hunk 13]
```diff
@@ -2768,7 +2768,7 @@
...
 			The anisotropic filtering level also affects decals and light projectors if they are configured to use anisotropic filtering. See [member rendering/textures/decals/filter] and [member rendering/textures/light_projectors/filter].
 			[b]Note:[/b] For performance reasons, anisotropic filtering [i]is not enabled by default[/i] on 2D and 3D materials. For this setting to have an effect in 3D, set [member BaseMaterial3D.texture_filter] to [constant BaseMaterial3D.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant BaseMaterial3D.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on materials. For this setting to have an effect in 2D, set [member CanvasItem.texture_filter] to [constant CanvasItem.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant CanvasItem.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on the [CanvasItem] node displaying the texture (or in [CanvasTexture]). However, anisotropic filtering is rarely useful in 2D, so only enable it for textures in 2D if it makes a meaningful visual difference.
```
Comment (Mickeon) [inline doc/classes/ProjectSettings.xml:2770]: It is with glamorous stupor that I realise just now that you did not write this section from scratch and borrowed it from this description. We may as change it then, for consistency. Apologies for not being thorough and providing a review all at once.
```suggestion
			[b]Note:[/b] In 3D, for this setting to have an effect, set [member BaseMaterial3D.texture_filter] to [constant BaseMaterial3D.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant BaseMaterial3D.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on materials.
			[b]Note:[/b] In 2D, for this setting to have an effect, set [member CanvasItem.texture_filter] to [constant CanvasItem.TEXTURE_FILTER_LINEAR_WITH_MIPMAPS_ANISOTROPIC] or [constant CanvasItem.TEXTURE_FILTER_NEAREST_WITH_MIPMAPS_ANISOTROPIC] on the [CanvasItem] node displaying the texture (or in [CanvasTexture]). However, anisotropic filtering is rarely useful in 2D, so only enable it for textures in 2D if it makes a meaningful visual difference.
```

**Review by Mickeon — APPROVED** [review]
Documentation is very nice, but just a few things.

File: doc/classes/ProjectSettings.xml (hunk 12 above)
Comment (wagnerfs) [reply doc/classes/ProjectSettings.xml:2771]: I feel like this one has to stay, it was suggested by Calinou but the word has been used a lot in the same file every time you must refer to another method to change something at runtime

Comment (Mickeon) [reply doc/classes/ProjectSettings.xml:2771]: Okay, in that case. If it's for consistency with the rest of the file. Perhaps it's worth a refresher some other day.

**Review by clayjohn — COMMENTED** [review]
The code looks great to me! Sorry, I had this on my backlog to review for several months. I didn't intend for it to take so long. 

One question I have is whether there is a benefit to specifying anisotropy per viewport. Previously when you changed the project setting it would change for all viewports. With this change, only the root viewport will be affected (which is technically a slight compatibility breakage). The trade off is that we can now control anisotropy on a per-viewport basis. But I'm not sure that doing so is desirable from a user point of view. 

Can anyone think of a situation where you would want to change anisotropy in one viewport but not another?

**Comment by mrjustaguy** [comment]
There are possible reasons to avoid using AF on other viewports, say if you're making camera feeds or planar reflections you may not benefit from the filtering due to how they're presented while you'd be paying the cost (albeit slight on today's hardware)

**Comment by Calinou** [comment]
> There are possible reasons to avoid using AF on other viewports, say if you're making camera feeds or planar reflections you may not benefit from the filtering due to how they're presented while you'd be paying the cost (albeit slight on today's hardware)

Drawing a real-time satellite minimap with a viewport will also not benefit from anisotropic filtering, since textures will always be viewed head-on in this situation.

**Comment by Mickeon** [comment]
I concur with Calinou. Ideally there should be an option to inherit the level from the project setting. Most users would not be willing to change this setting for every Viewport to suit their needs.

**Comment by wagnerfs** [comment]
The code -as is- right now inherits the value from the project settings when the subviewport is added, so loading a project that had no settings on it prior to this PR would set all viewports to current project value just fine.
I -could- add an extra enum along with the anisotropy levels and make it default that would make subviewports default to project settings, should work out of the box and prevent breaking compatibility.

Would be nice to have a concrete decision and direction about this feature though as it's been almost a whole year and I don't really have the same free time I had early this year.

**Comment by mrjustaguy** [comment]
IMO what Calinou said this should be done for other parameters like MSAA, so is outside of the scope of this PR

**Comment by Mickeon** [comment]
Good point, then this could still use a rebase given the conflict.

**Comment by wagnerfs** [comment]
Should be up to date now!

**Review by clayjohn — APPROVED** [review]
Approving. The code looks great!

I am still skeptical about the need to expose this on a per-viewport basis. I am especially wary that the main reason this PR exposes anisotropy on a per-viewport basis was simply because it was easier than the alternative (because it can piggyback on our mip offset code). However, enough people have commented on the PR with reasons why they might need this exposed on a per viewport basis that I won't fight against it. 

Overall, the code additions are small and this won't add much maintenance burden beyond what we have already accepted for the mip level offset

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:
